error: Expected TypedDict key to be string literal for `Union` of `TypedDict` and `dict`

<!--
If you're not sure whether what you're experiencing is a mypy bug, please see the "Question and Help" form instead.
Please consider:

- checking our common issues page: https://mypy.readthedocs.io/en/stable/common_issues.html
- searching our issue tracker: https://github.com/python/mypy/issues to see if it's already been reported
- asking on gitter chat: https://gitter.im/python/typing
-->

**Bug Report**

<!--
If you're reporting a problem with a specific library function, the typeshed tracker is better suited for this report: https://github.com/python/typeshed/issues

If the project you encountered the issue in is open source, please provide a link to the project.
-->

Error for a valid input to `TypedDict | dict[int, float]`

**To Reproduce**
https://mypy-play.net/?gist=3d0a2913c33c47b405cb297df1607e8f

```python
from typing import TypedDict

class Foo(TypedDict):
    some: str
    name: str

type Bar = dict[int, float]

type Baz = Foo | Bar

baz: Baz = {"some": "foo", "name": "x"}  # OK
bar: Bar = {1: 5.2, 2: 7.4}  # OK

baz2: Baz = {1: 5.2} # error: Expected TypedDict key to be string literal  [misc]
```

**Expected Behavior**
I would expect `baz2` to be valid

**Actual Behavior**

```
main.py:14: error: Expected TypedDict key to be string literal  [misc]
Found 1 error in 1 file (checked 1 source file)
```

When mis-typing the dict variant mypy does give an expected error:
```python
baz3: Baz = {"1": 5.2} # error: Dict entry 0 has incompatible type "str": "float"; expected "int": "float"  [dict-item]
```

**Your Environment**

<!-- Include as many relevant details about the environment you experienced the bug in -->

- Mypy version used: 2.1.0

Original discussion: https://github.com/python/typing/discussions/2292

<!-- You can freely edit this text, please remove all the lines you believe are unnecessary. -->

--- a/mypy/checkexpr.py
+++ b/mypy/checkexpr.py
@@ -932,7 +932,8 @@ class ExpressionChecker(ExpressionVisitor[Type], ExpressionCheckerSharedApi):
         kwargs: list[tuple[Expression | None, Expression]],
         context: Context,
     ) -> bool:
-        result = self.validate_typeddict_kwargs(kwargs=kwargs, callee=callee)
+        with self.msg.filter_errors():
+            result = self.validate_typeddict_kwargs(kwargs=kwargs, callee=callee)
         if result is not None:
             validated_kwargs, _ = result
             return callee.required_keys <= set(validated_kwargs.keys()) <= set(callee.items.keys())

--- a/test-data/unit/check-typeddict.test
+++ b/test-data/unit/check-typeddict.test
@@ -941,6 +941,18 @@ c: Union[A, B] = {'@type': 'a-type', 'value': 'Test'}  # E: Type of TypedDict is
 [builtins fixtures/dict.pyi]
 [typing fixtures/typing-typeddict.pyi]
 
+[case testTypedDictUnionWithNonTypedDictLiteralKey]
+from typing import Union, TypedDict
+
+class Foo(TypedDict):
+    some: str
+    name: str
+
+baz: Union[Foo, dict[int, float]] = {1: 5.2}
+baz2: Union[Foo, dict[int, float]] = {'some': 'x', 'name': 'y'}
+[builtins fixtures/dict.pyi]
+[typing fixtures/typing-typeddict.pyi]
+
 -- Use dict literals
 
 [case testTypedDictDictLiterals]

